proj/hps_accel/src/tensorflow/lite/kernels/internal/reference/integer_ops/conv_accel.cc

```diff
@@ -117,22 +120,20 @@ void ConvPerChannel4x4(const ConvParams& params,
         for (int out_channel = out_channel_offset;
              out_channel < out_channel_offset + output_channels;
              ++out_channel) {
-                         int iterations = filter_height * filter_width * input_depth / 16;
+          int iterations = filter_height * filter_width * input_depth / 16;
           hps_accel::AdvanceFilterInput(iterations);
           int32_t acc = multiply_accumulate();
           hps_accel::PostProcess(acc);
         }
 
-        // Extract outputs and send to memory
-        for (int out_channel = out_channel_offset;
-             out_channel < out_channel_offset + output_channels;
-             out_channel += 4) {
-          // TODO: less pointer arithmetic
-          uint32_t* output_data_ptr = static_cast<uint32_t*>(
-              static_cast<void*>(output_data + Offset(output_shape, 0, out_y,
-                                                      out_x, out_channel)));
-          *output_data_ptr = hps_accel::GetOutputWord();
+        // Pull result data from output FIFO and place into memory, a word at a
+        // time.
+        uint32_t* output_data32 = output_data32_base;
```

Comment:
I can see how this is a useful optimisation (does it produce a substantial improvement in speed by itself, or is it really just preparation for doing some fancier DMA later?).

But I liked the safety of Offset() doing the pointer arithematic for me, to reduce the likelihood I would get some dimension wrong somewhere. So I'm a bit sad we lose that.

Would it be worth adding a DCHECK inside the loop which cross-checks the pointer against the expected offset? Or I am being too paranoid perhaps?

Reply:
Offset() is rather slow, so moving it outside the loops seems best.

The "safety" of Offset() is rather dubious when DCHECKs are off.

I'll add a DCHECK in a follow up PR.

Reply:
Yes it's only safety in the sense that when I come along later and try to change this code, I would be able to turn off NDEBUG and satisfy myself I haven't messed up the pointer arithmetic.